File: windows/Services/SyncMerge.cs
using Hatch.Models;

namespace Hatch.Services;

// Record-level (not field-level) last-write-wins union of two TasksFiles by Id.
// A task/list edited on only one side is kept as-is; a task/list present on both
// sides keeps whichever copy has the later UpdatedAt. Nothing is ever dropped —
// this is the non-destructive alternative to "pick a side, lose the other".
public static class SyncMerge
{
    public static TasksFile Merge(TasksFile local, TasksFile server) => new()
    {
        Tasks = MergeById(local.Tasks, server.Tasks, t => t.Id, t => t.UpdatedAt),
        Lists = MergeById(local.Lists, server.Lists, l => l.Id, l => l.UpdatedAt)
    };

    private static List<T> MergeById<T>(
        List<T> local, List<T> server, Func<T, Guid> idOf, Func<T, DateTimeOffset> updatedAtOf)
    {
        var merged = new Dictionary<Guid, T>();
        foreach (var item in server)
            merged[idOf(item)] = item;

        foreach (var item in local)
        {
            var id = idOf(item);
            if (!merged.TryGetValue(id, out var existing) || updatedAtOf(item) >= updatedAtOf(existing))
                merged[id] = item;
        }

        return [.. merged.Values];
    }
}

File: mobile/shared/src/commonMain/kotlin/dev/hatch/sync/SyncMerge.kt
package dev.hatch.sync

import kotlin.time.Instant

// Transcription of windows/Services/SyncMerge.cs and docs/sync-protocol.md §5.
// Record-level last-write-wins union keyed by Id; local wins an exact tie.
object SyncMerge {

    fun merge(local: TasksFile, server: TasksFile): TasksFile = TasksFile(
        tasks = mergeById(local.tasks, server.tasks, { it.id }, { it.updatedAt }),
        lists = mergeById(local.lists, server.lists, { it.id }, { it.updatedAt }),
    )

    private fun <T> mergeById(
        local: List<T>,
        server: List<T>,
        idOf: (T) -> String,
        updatedAtOf: (T) -> String,
    ): List<T> {
        val merged = LinkedHashMap<String, T>()
        for (item in server) merged[idOf(item)] = item

        for (item in local) {
            val id = idOf(item)
            val existing = merged[id]
            if (existing == null || instantOf(updatedAtOf(item)) >= instantOf(updatedAtOf(existing))) {
                merged[id] = item
            }
        }
        return merged.values.toList()
    }

    // Unparseable must lose: treating it as "now" would let bad data overwrite good.
    private fun instantOf(text: String): Instant =
        runCatching { Instant.parse(text) }.getOrElse { Instant.DISTANT_PAST }
}
